File: backend/routes/alquileres/cotizacion.py
"""Cotización canónica del carrito (#501 — extraído del god-module `routes/alquileres.py`).

Fuente única del total del carrito: el front manda solo `items` (equipo_id +
cantidad) y fechas; el backend pone los precios (de `equipos`), el perfil/descuento
del cliente y devuelve el desglose de `services.precios.calcular_total`. Registra su
ruta sobre el router compartido del paquete `routes.alquileres`.
"""
from typing import Optional

from fastapi import Request
from pydantic import BaseModel, field_validator

from database import get_db, to_datetime
from rate_limit import limiter
from auth.guards import is_admin_email
from auth.session import dev_bypass_enabled, get_session
from services.precios import (
    bruto_linea,
    calcular_total,
    jornadas_periodo,
    precio_jornada_efectivo,
    tipos_equipo_batch,
)
from descuentos.queries.decision import resolver_origen_pedido_monto
from descuentos.queries.jornadas import obtener_descuento_jornadas
from routes.alquileres.core import router, _resolver_descuentos_snapshot_o_vivo
# Validadores de descuento: fuente única compartida con `PedidoDatos`
# (routes/alquileres/modelos.py) — antes estaban duplicados byte a byte acá.
from routes.alquileres.modelos import (
    _validar_descuento_manual_monto,
    _validar_descuento_manual_tipo,
    _validar_descuento_pct,
)


class CotizarItem(BaseModel):
    # equipo_id None = línea personalizada (#805): su precio y modo de cobro
    # vienen del front (el admin la edita libre); no se busca en `equipos`.
    equipo_id: Optional[int] = None
    cantidad: int
    precio_jornada: Optional[int] = None
    cobro_modo: Optional[str] = None


class CotizarRequest(BaseModel):
    items: list[CotizarItem] = []
    fecha_desde: Optional[str] = None
    fecha_hasta: Optional[str] = None
    # Los dos siguientes SOLO los honra una sesión admin (el builder de pedidos
    # arma para OTRO cliente, no para la sesión):
    #  - cliente_id: de qué cliente tomar el perfil tributario.
    #  - descuento_pct: override del descuento del cliente (el admin lo edita
    #    en vivo en el builder; gana sobre el `clientes.descuento` guardado).
    cliente_id: Optional[int] = None
    descuento_pct: Optional[float] = None
    # Override manual en % o en $ fijo (Fase C-2, #1219): mismo par que
    # `PedidoDatos` (routes/alquileres/core.py) — el builder los edita en vivo
    # para que el preview coincida con lo que se va a persistir.
    descuento_manual_tipo: Optional[str] = None
    descuento_manual_monto: Optional[float] = None
    # Solo lo honra una sesión admin: respeta el `precio_jornada` que manda cada
    # ítem de catálogo (el snapshot congelado del pedido que se está editando)
    # en vez de re-buscarlo en `equipos`. Sin esto, el editor de pedidos admin
    # mostraba un total "en vivo" con el precio de HOY del catálogo — distinto
    # al que persiste `_recalcular_total_pedido` al guardar (que sí respeta el
    # precio de línea) → dos totales del mismo pedido que podían no coincidir.
    # Ver MEMORIA 2026-06-06 "Datos del pedido: plata congelada".
    respetar_precio_item: Optional[bool] = False
    # Solo lo honra una sesión admin: id del pedido que se está editando. Cuando
    # el pedido ya NO está en `presupuesto` (plata congelada), el descuento de
    # cliente/jornadas del preview sale del MISMO snapshot que la persistencia
    # (`_resolver_descuentos_snapshot_o_vivo`), no en vivo — así el total del
    # editor coincide con `monto_total` (y con la lista de pedidos). El editor
    # solo lo manda para pedidos no-presupuesto; en presupuesto el descuento
    # sigue al cliente en vivo. Ver MEMORIA 2026-06-06 "plata congelada".
    pedido_id: Optional[int] = None
    # #1240: a nombre de quién se está cotizando (perfil personal alternativo o
    # productora) — solo lo honra una sesión cliente (mismo criterio que el resto
    # de este bloque: el admin cotiza para el cliente del pedido, no para sí
    # mismo). Mutuamente excluyentes; NULL/NULL = perfil default de la cuenta.
    perfil_fiscal_id: Optional[int] = None
    productora_id: Optional[int] = None

    # Mismo validador que `PedidoDatos.descuento_pct` (routes/alquileres/modelos.py)
    # — este override vivía sin cota de rango (hallazgo de la Fase A del split
    # de descuentos/, #1184): un admin podía mandar un negativo al preview en
    # vivo sin que nada lo rechazara. Ahora es la MISMA función (no una copia):
    # preview y guardado no pueden divergir en qué rechazan.
    @field_validator("descuento_pct")
    @classmethod
    def validate_descuento(cls, v):
        return _validar_descuento_pct(v)

    @field_validator("descuento_manual_tipo")
    @classmethod
    def validate_descuento_manual_tipo(cls, v):
        return _validar_descuento_manual_tipo(v)

    @field_validator("descuento_manual_monto")
    @classmethod
    def validate_descuento_manual_monto(cls, v):
        return _validar_descuento_manual_monto(v)


@router.post("/cotizar")
@limiter.limit("30/minute")
def cotizar(data: CotizarRequest, request: Request):
    """Cotización canónica del carrito — fuente única, calculada en el backend.

    El front NO calcula el total: manda solo `items` (equipo_id + cantidad) y,
    si las hay, las fechas. El backend pone TODO lo demás:
    - el `precio_jornada` de cada equipo (de la tabla `equipos`, no se confía
      en lo que mande el front),
    - el perfil tributario y el descuento del cliente (cliente logueado → el
      suyo; admin → el del `cliente_id` que pase; anónimo → `consumidor_final`),
    - el descuento por jornadas.

    **Sin fechas = modo estimado:** jornadas=1, sin descuentos ni IVA (es solo
    referencia de precio mientras el cliente no eligió período). Replica el
    comportamiento histórico del carrito en armado.

    Devuelve el desglose de `services.precios.calcular_total` + `descuento_origen`
    y `subtotal_por_jornada` (para el UI), de modo que el front muestre los
    números tal cual. Reemplaza el cálculo duplicado del front
    (`src/lib/cart-total.ts`). Ver #617.
    """
    with get_db() as conn:
        # Jornadas: misma fórmula única (ceil/24h). Sin fechas → 1.
        d0 = to_datetime(data.fecha_desde) if data.fecha_desde else None
        d1 = to_datetime(data.fecha_hasta) if data.fecha_hasta else None
        jornadas = jornadas_periodo(d0, d1)
        tiene_fechas = bool(d0 and d1)

        session = get_session(request)
        # Mismo criterio que `require_admin` (auth/guards.py): el bypass de
        # dev (ADMIN_BYPASS_AUTH=1, nunca en prod) también cuenta como admin
        # acá — si no, el preview en vivo ignora el override de descuento
        # aunque el guardado (que sí pasa por require_admin) lo haya aceptado.
        es_admin = dev_bypass_enabled() or bool(session and is_admin_email(session.get("email")))
        respetar_precio_item = es_admin and bool(data.respetar_precio_item)

        # Precios desde el backend. Equipos inexistentes/eliminados se ignoran
        # (cotización best-effort: el carrito puede tener algo que ya no está).
        # Fetch por-ítem (lookup por PK indexada): se revirtió el batch `IN (...)`
        # de #643 que devolvía precios_map vacío en prod → total $0 (regresión).
        # `tipos_equipo_batch` (solo id+tipo, query aparte) SÍ va en batch — no
        # toca el precio, resuelve `es_combo` para el descuento no-acumulable
        # (Fase C-3, #1219).
        tipos = tipos_equipo_batch(
            conn, [it.equipo_id for it in data.items if it.equipo_id is not None]
        )
        items_para_total = []
        for it in data.items:
            if it.cantidad <= 0:
                continue
            # Línea personalizada (#805): no es del catálogo → su precio y modo de
            # cobro los pone el front (el admin la edita libre, igual que ya confía
            # el precio editable por línea en PUT /items). No reserva stock.
            if it.equipo_id is None:
                items_para_total.append({
                    "equipo_id": None,
                    "cantidad": it.cantidad,
                    "precio_jornada": int(it.precio_jornada or 0),
                    "cobro_modo": it.cobro_modo or "jornada",
                })
                continue
            # Admin editando un pedido existente (`respetar_precio_item`): usa el
            # precio de línea que manda el front (el snapshot congelado del
            # pedido, editable por el admin) en vez de recotizar contra el
            # catálogo — así el total "en vivo" del editor coincide siempre con
            # el que persiste `_recalcular_total_pedido` al guardar.
            if respetar_precio_item and it.precio_jornada is not None:
                precio = int(it.precio_jornada)
            else:
                # Precio efectivo por jornada (combo → derivado de componentes
                # C3 #635; kit/simple → su precio propio), fuente única.
                precio = precio_jornada_efectivo(conn, it.equipo_id)
                if precio is None:
                    continue
            items_para_total.append({
                "equipo_id": it.equipo_id,
                "cantidad": it.cantidad,
                "precio_jornada": precio,
                "cobro_modo": "jornada",
                "es_combo": tipos.get(it.equipo_id) == "combo",
            })
        subtotal_por_jornada = sum(
            it["precio_jornada"] * it["cantidad"] for it in items_para_total
        )

        # Perfil tributario + descuento del cliente. Solo en modo firme (con
        # fechas): sin fechas es un estimado sin IVA ni descuentos.
        perfil = None
        descuento_cliente_pct = 0.0
        descuento_jornadas_pct = 0.0
        descuento_manual_pct = 0.0
        descuento_manual_tipo = "pct"
        descuento_manual_monto = 0.0
        if tiene_fechas:
            # ¿El preview es el editor de un pedido con la plata YA congelada?
            # Entonces el descuento (cliente + jornadas) sale del MISMO snapshot
            # que persiste el guardado (`_resolver_descuentos_snapshot_o_vivo`),
            # NO en vivo — si no, el total del editor mostraría un descuento
            # distinto al de `monto_total` / la lista de pedidos (MEMORIA
            # 2026-06-06 "plata congelada"). El perfil fiscal (IVA) sí sigue en
            # vivo. Presupuesto (o sin `pedido_id`) → flujo en vivo de siempre.
            pedido_congelado = None
            if es_admin and data.pedido_id:
                pedido_congelado = conn.execute(
                    "SELECT estado, cliente_id, descuento_jornadas_pct, "
                    "descuento_cliente_pct FROM alquileres WHERE id=%s",
                    (data.pedido_id,),
                ).fetchone()
                if pedido_congelado and pedido_congelado["estado"] == "presupuesto":
                    pedido_congelado = None  # presupuesto sigue el descuento en vivo
            # ¿Para qué cliente se cotiza?
            #   - Admin (back-office): SIEMPRE el cliente del pedido (`data.cliente_id`),
            #     nunca la ficha de la propia sesión. La admin-ness la da el EMAIL
            #     (`require_admin`/`is_admin_email`), no el `role` — un dueño que se
            #     logueó por el portal de cliente tiene una sesión `role="cliente"` +
            #     su propio `cliente_id` Y además es admin. Con la precedencia vieja
            #     (rama `role=="cliente"` primero) el builder cotizaba con el descuento
            #     de la ficha del PROPIO admin para TODOS los pedidos, no la del pedido
            #     → "descuento fantasma" (#1231). El camino que PERSISTE la plata
            #     (`_recalcular_total_pedido`) ya usaba el cliente del pedido, así que
            #     esto era solo el preview mintiendo, no datos corruptos.
            #   - Cliente puro (portal): su propia ficha.
            #   - Anónimo (sin sesión): sin descuento de cliente.
            target_cliente_id = None
            es_sesion_cliente = False
            if es_admin:
                target_cliente_id = data.cliente_id
            elif session and session.get("role") == "cliente" and session.get("cliente_id"):
                target_cliente_id = session["cliente_id"]
                es_sesion_cliente = True
            if target_cliente_id:
                c = conn.execute(
                    "SELECT perfil_impuestos, descuento FROM clientes WHERE id=%s",
                    (target_cliente_id,),
                ).fetchone()
                if c:
                    perfil = c["perfil_impuestos"]
                    # Congelado: el descuento sale del snapshot (más abajo), no
                    # del vivo. El perfil (IVA) sí es en vivo siempre.
                    if pedido_congelado is None:
                        descuento_cliente_pct = c["descuento"] or 0.0
                # #1240: solo la sesión cliente puede elegir facturar a nombre de
                # un perfil personal alternativo o una productora (el admin no
                # manda estos campos para el pedido de otro cliente).
                # `_resolver_datos_fiscales_pedido` scopea `perfil_fiscal_id` por
                # `cliente_id` sola, pero NO valida membership de `productora_id`
                # (asume que el caller ya lo hizo, como sí hace la creación real
                # del pedido) — acá se valida explícitamente antes de usarlo, para
                # no dejar que cualquier sesión cliente cotice con el perfil
                # fiscal de una productora ajena solo adivinando su id.
                productora_id_valida = None
                if es_sesion_cliente and data.productora_id:
                    vinculado = conn.execute(
                        "SELECT 1 FROM productora_miembros WHERE productora_id = %s AND cliente_id = %s",
                        (data.productora_id, target_cliente_id),
                    ).fetchone()
                    if vinculado:
                        productora_id_valida = data.productora_id
                if es_sesion_cliente and (data.perfil_fiscal_id or productora_id_valida):
                    from services.pedidos_enriquecimiento import _resolver_datos_fiscales_pedido

                    fiscal = _resolver_datos_fiscales_pedido(
                        conn, target_cliente_id, data.perfil_fiscal_id, productora_id_valida
                    )
                    if fiscal.get("perfil_impuestos"):
                        perfil = fiscal["perfil_impuestos"]
            # Override MANUAL del admin (lo edita en vivo en el builder del
            # pedido) — jerarquía (Fase C-1, #1219): gana OUTRIGHT sobre
            # cliente/jornadas, ya NO reemplaza el descuento del cliente para
            # que compita por tamaño. Fase C-2 (#1219): el override puede ser
            # % (de siempre) o $ fijo, mismo campo de la UI.
            if es_admin and data.descuento_pct:
                descuento_manual_pct = data.descuento_pct
            if es_admin and data.descuento_manual_tipo:
                descuento_manual_tipo = data.descuento_manual_tipo
            if es_admin and data.descuento_manual_monto:
                descuento_manual_monto = data.descuento_manual_monto
            if pedido_congelado is None:
                descuento_jornadas_pct = obtener_descuento_jornadas(conn, jornadas)
            else:
                # Pedido no-presupuesto: descuento (jornadas + cliente) del MISMO
                # snapshot que la persistencia → el editor no puede divergir de
                # `monto_total`. El override manual (arriba) sí sigue en vivo.
                descuento_jornadas_pct, descuento_cliente_pct = (
                    _resolver_descuentos_snapshot_o_vivo(conn, pedido_congelado, jornadas)
                )

        desglose = calcular_total(
            items=items_para_total,
            jornadas=jornadas,
            descuento_cliente_pct=descuento_cliente_pct,
            descuento_jornadas_pct=descuento_jornadas_pct,
            descuento_manual_pct=descuento_manual_pct,
            descuento_manual_tipo=descuento_manual_tipo,
            descuento_manual_monto=descuento_manual_monto,
            perfil_impuestos=perfil,
        )

        # Cuál descuento ganó (para el label del UI) — misma jerarquía que
        # decidió el pct en `calcular_total`, así nunca puede divergir.
        descuento_origen = resolver_origen_pedido_monto(
            descuento_manual_tipo, descuento_manual_pct, descuento_manual_monto,
            descuento_cliente_pct, descuento_jornadas_pct,
        )

        # Desglose POR LÍNEA para que el front MUESTRE (no calcule) el detalle por
        # equipo: `subtotal_por_jornada` (siempre, el "$X/día" del ítem) + `bruto`/`neto`
        # del período (cuando hay fechas). El neto por línea reparte el descuento ganado
        # con el mismo redondeo por línea que usaba el front; la suma de netos por línea
        # puede diferir del neto total por unos pesos (redondeo independiente por línea)
        # → el TOTAL autoritativo es `neto` (top-level), las líneas son detalle de display.
        # `equipo_id` None = línea personalizada (#805). Combos (`es_combo`, Fase
        # C-3 #1219) no reciben el % ganador — ya vienen con su propio descuento
        # de componente horneado en `precio_jornada`.
        pct_aplicado = desglose["descuento_pct"]
        lineas = []
        for it in items_para_total:
            bruto = bruto_linea(it, jornadas)
            dto = 0 if it.get("es_combo") else int(round(bruto * pct_aplicado / 100))
            lineas.append({
                "equipo_id": it["equipo_id"],
                "cantidad": it["cantidad"],
                "precio_jornada": it["precio_jornada"],
                "subtotal_por_jornada": it["precio_jornada"] * it["cantidad"],
                "bruto": bruto,
                "neto": bruto - dto,
            })

        return {
            "jornadas": jornadas,
            "subtotal_por_jornada": int(subtotal_por_jornada),
            "descuento_origen": descuento_origen,
            "lineas": lineas,
            **desglose,
        }

File: frontend/src/lib/cotizacion.ts
/**
 * cotizacion.ts — Cliente del endpoint canónico de cotización (`POST /api/cotizar`).
 *
 * El total del carrito lo calcula EL BACKEND (fuente única,
 * `services/precios.calcular_total`). El front ya NO reimplementa la fórmula:
 * manda los ítems (equipo_id + cantidad) y, si hay, las fechas, y muestra el
 * desglose que devuelve el backend tal cual. Reemplaza al viejo `cart-total.ts`.
 * Ver #617 / #508.
 *
 * Reglas (las decide el backend, no el front):
 *  - Sin fechas → estimado de UNA jornada, sin descuento ni IVA.
 *  - Cliente logueado → su perfil/descuento. Admin → el del `clienteId` que
 *    pase (arma pedidos para terceros). Anónimo / sin cliente → consumidor final.
 */

import { useEffect, useRef, useState } from "react";
import { useQuery, keepPreviousData } from "@tanstack/react-query";
import { authedPostJson } from "@/lib/authedFetch";

/** Debounce del recálculo en vivo. Sin esto, `useCotizacion` le pega a
 * `/api/cotizar` en CADA tecla (tipear un %, arrastrar una fecha, cambiar una
 * cantidad) → editando activamente se supera el rate-limit del endpoint
 * (30/min) → 429 → el desglose se queda mostrando el último valor bueno en
 * silencio (bug "el descuento no se aplica"). Coalescemos los cambios rápidos
 * en UN solo request. El primer cálculo NO se debouncea (estado inicial
 * inmediato). */
const COTIZAR_DEBOUNCE_MS = 300;

export type DescuentoOrigen = "manual" | "cliente" | "jornadas" | "ninguno";

/** Detalle por equipo que el front MUESTRA (no calcula): el backend lo resuelve
 *  en `/api/cotizar` con el precio efectivo (combo-aware). Ver FASE 3 / #1110. */
export type CotizacionLinea = {
  /** Backend id del equipo (null = línea personalizada del builder admin, #805). */
  equipoId: number | null;
  cantidad: number;
  /** Precio efectivo por jornada del ítem (combo → derivado de componentes). */
  precioJornada: number;
  /** precioJornada × cantidad (el "$X/día" del ítem). */
  subtotalPorJornada: number;
  /** Bruto del período: × jornadas (antes de descuento). */
  bruto: number;
  /** Neto del período: con el descuento ganado repartido por línea. */
  neto: number;
};

/** Desglose normalizado para el UI (mismas keys que usaba `CartTotal`). */
export type Cotizacion = {
  /** Σ(precio_jornada × cantidad), sin multiplicar por jornadas. */
  subtotalPorJornada: number;
  jornadas: number;
  /** Bruto del período: subtotalPorJornada × jornadas (antes de descuentos). */
  subtotal: number;
  /** Bruto SIN las líneas de combo (C-3, #1219) — el tope real de un
   *  descuento manual en $: no se le puede descontar más a un pedido que a
   *  su parte descontable (los combos ya vienen con su propio descuento). */
  subtotalDescontable: number;
  descuentoPct: number;
  descuentoOrigen: DescuentoOrigen;
  descuentoMonto: number;
  /** Neto = subtotal − descuento. Es lo que se PERSISTE en `monto_total`. */
  totalNeto: number;
  /** IVA (0 si el cliente no es responsable_inscripto). */
  iva: number;
  conIva: boolean;
  /** Total a mostrar: neto + IVA. */
  total: number;
  /** Detalle por equipo (para el caján/teasers; el front lo muestra, no lo calcula). */
  lineas: CotizacionLinea[];
};

/** Una línea cruda del backend. */
type CotizarLineaResp = {
  equipo_id: number | null;
  cantidad: number;
  precio_jornada: number;
  subtotal_por_jornada: number;
  bruto: number;
  neto: number;
};

/** Respuesta cruda del backend (`/api/cotizar`). */
type CotizarResp = {
  jornadas: number;
  subtotal_por_jornada: number;
  descuento_origen: DescuentoOrigen;
  bruto: number;
  bruto_descontable: number;
  descuento_pct: number;
  descuento_monto: number;
  neto: number;
  con_iva: boolean;
  iva_pct: number;
  iva_monto: number;
  total_final: number;
  lineas?: CotizarLineaResp[];
};

export const COTIZACION_VACIA: Cotizacion = {
  subtotalPorJornada: 0,
  jornadas: 1,
  subtotal: 0,
  subtotalDescontable: 0,
  descuentoPct: 0,
  descuentoOrigen: "ninguno",
  descuentoMonto: 0,
  totalNeto: 0,
  iva: 0,
  conIva: false,
  total: 0,
  lineas: [],
};

function adaptar(r: CotizarResp): Cotizacion {
  return {
    subtotalPorJornada: r.subtotal_por_jornada,
    jornadas: r.jornadas,
    subtotal: r.bruto,
    subtotalDescontable: r.bruto_descontable,
    descuentoPct: r.descuento_pct,
    descuentoOrigen: r.descuento_origen,
    descuentoMonto: r.descuento_monto,
    totalNeto: r.neto,
    iva: r.iva_monto,
    conIva: r.con_iva,
    total: r.total_final,
    lineas: (r.lineas ?? []).map((l) => ({
      equipoId: l.equipo_id,
      cantidad: l.cantidad,
      precioJornada: l.precio_jornada,
      subtotalPorJornada: l.subtotal_por_jornada,
      bruto: l.bruto,
      neto: l.neto,
    })),
  };
}

/** Lookup por equipo_id (backend id) del detalle por línea de una cotización. */
export function lineaPorEquipo(
  cot: Cotizacion,
  equipoId: number | null | undefined,
): CotizacionLinea | undefined {
  if (equipoId == null) return undefined;
  return cot.lineas.find((l) => l.equipoId === equipoId);
}

export type CotizarItemInput = {
  /** null = línea personalizada (#805): manda su precio y modo (no se busca en `equipos`). */
  equipoId: number | null;
  cantidad: number;
  /** Línea personalizada: precio libre y modo de cobro. */
  precioJornada?: number;
  cobroModo?: "jornada" | "fijo";
};

/**
 * Cotiza el carrito contra el backend. Refetch automático al cambiar ítems o
 * fechas; mantiene el valor anterior mientras recalcula (sin parpadeo).
 * Carrito vacío → devuelve `COTIZACION_VACIA` sin pegarle al backend.
 */
export function useCotizacion(args: {
  items: CotizarItemInput[];
  /** ISO-local `YYYY-MM-DDTHH:MM:00` (usar `toLocalISO`). Sin fechas → estimado. */
  fechaDesde?: string | null;
  fechaHasta?: string | null;
  /** Solo admin: cotizar para el cliente del pedido. */
  clienteId?: number | null;
  /** Solo admin: override manual del pedido en % (el builder lo edita en vivo). */
  descuentoPct?: number | null;
  /** Solo admin, Fase C-2 (#1219): tipo del override manual — "pct" (default)
   *  o "monto" ($ fijo). Mismo campo de la UI, un selector al lado. */
  descuentoTipo?: "pct" | "monto" | null;
  /** Solo admin, Fase C-2: override manual del pedido en $ fijo, usado cuando
   *  `descuentoTipo === "monto"`. */
  descuentoMonto?: number | null;
  /**
   * Solo admin, para el editor de un pedido YA EXISTENTE: respeta el
   * `precioJornada` de cada ítem (el snapshot congelado del pedido, editable
   * por el admin) en vez de recotizar contra el precio de catálogo de HOY.
   * Sin esto, el total "en vivo" del editor podía no coincidir con el que
   * persiste el guardado — dos totales del mismo pedido. Ver MEMORIA
   * 2026-06-06 "Datos del pedido: plata congelada".
   */
  respetarPrecioItem?: boolean;
  /**
   * Identidad de la entidad DUEÑA de esta cotización (ej. el id del pedido).
   * Sin esto, el hook asume una identidad estable (ej. "el carrito"), donde
   * mostrar el valor anterior mientras se recalcula es lo deseado (sin
   * parpadeo al editar). Con un `resetKey` que cambia — ej. el admin navega
   * a OTRO pedido en un panel maestro-detalle que no desmonta el componente
   * (`/admin/pedidos/$id`) — el hook deja de mostrar el placeholder de la
   * entidad ANTERIOR: mejor un instante en $0 que el desglose de otro pedido
   * superpuesto al nombre del nuevo (bug real, confirmado con los valores en
   * 0 en la base — no era un cálculo mal hecho, era este hook mostrando la
   * cotización vieja mientras la nueva todavía viajaba por la red).
   */
  resetKey?: string | number | null;
  enabled?: boolean;
}): { data: Cotizacion; isFetching: boolean; isError: boolean } {
  const {
    items,
    fechaDesde,
    fechaHasta,
    clienteId,
    descuentoPct,
    descuentoTipo,
    descuentoMonto,
    respetarPrecioItem = false,
    resetKey = null,
    enabled = true,
  } = args;

  const body = {
    items: items
      .filter((i) => i.cantidad > 0)
      .map((i) => ({
        equipo_id: i.equipoId,
        cantidad: i.cantidad,
        // Personalizadas (equipoId null): siempre mandan precio/modo propios.
        // Catálogo con `respetarPrecioItem`: el backend solo lo honra si la
        // sesión es admin; en el resto de los usos (carrito público) el
        // backend ignora estos campos y toma el precio de `equipos`.
        ...(i.equipoId == null || respetarPrecioItem
          ? { precio_jornada: i.precioJornada ?? 0, cobro_modo: i.cobroModo ?? "jornada" }
          : {}),
      })),
    fecha_desde: fechaDesde ?? null,
    fecha_hasta: fechaHasta ?? null,
    cliente_id: clienteId ?? null,
    descuento_pct: descuentoPct ?? null,
    descuento_manual_tipo: descuentoTipo ?? null,
    descuento_manual_monto: descuentoMonto ?? null,
    respetar_precio_item: respetarPrecioItem,
  };

  // Debounce: la query usa el body DEBOUNCEADO. Al tipear rápido, `bodyKey`
  // cambia por tecla pero solo el último (tras COTIZAR_DEBOUNCE_MS de quietud)
  // dispara un fetch → un request por edición, no por tecla. El estado inicial
  // es inmediato (`useState(body)`).
  const bodyKey = JSON.stringify(body);
  const [debounced, setDebounced] = useState({ key: bodyKey, body });

  // Generación: se incrementa cada vez que cambia `resetKey`. Ajuste de
  // estado DURANTE el render (patrón oficial de React para resetear estado
  // al cambiar de identidad, no en un efecto) — salta el debounce en el
  // mismo render, antes de pintar, para que el body ya sea el de la entidad
  // nueva en el primer frame.
  const prevResetKeyRef = useRef(resetKey);
  const [generation, setGeneration] = useState(0);
  if (prevResetKeyRef.current !== resetKey) {
    prevResetKeyRef.current = resetKey;
    setGeneration((g) => g + 1);
    if (bodyKey !== debounced.key) setDebounced({ key: bodyKey, body });
  }

  useEffect(() => {
    if (bodyKey === debounced.key) return;
    const t = setTimeout(() => setDebounced({ key: bodyKey, body }), COTIZAR_DEBOUNCE_MS);
    return () => clearTimeout(t);
    // eslint-disable-next-line react-hooks/exhaustive-deps -- debounce: dispara por el hash del body; `body` se captura fresco cuando cambia el hash
  }, [bodyKey, debounced.key]);

  const hayItems = debounced.body.items.length > 0;
  const q = useQuery({
    queryKey: ["cotizar", debounced.body],
    queryFn: () => authedPostJson<CotizarResp>("/api/cotizar", debounced.body),
    enabled: enabled && hayItems,
    placeholderData: keepPreviousData,
    staleTime: 30_000,
  });

  // Mientras el dato que devuelve `q` sea un PLACEHOLDER (de `keepPreviousData`,
  // de una generación anterior) no lo mostramos como si fuera de la entidad
  // actual — se corrige solo apenas responde el fetch real de la generación
  // nueva. Sin `resetKey` (default `null`, nunca cambia) esto es un no-op:
  // mismo comportamiento de siempre para los demás consumidores del hook.
  const lastGoodGenerationRef = useRef(-1);
  if (q.data && !q.isPlaceholderData) {
    lastGoodGenerationRef.current = generation;
  }
  const identityPending = lastGoodGenerationRef.current !== generation;

  // `isFetching` cubre TAMBIÉN el intervalo de debounce (hay una edición
  // pendiente de recalcular) → el consumidor puede mostrar "recalculando…" y
  // no presentar el número viejo como definitivo. `isError` surfacea un fallo
  // del recálculo (ej. 429) en vez de quedarse mudo con el valor stale.
  const pendingDebounce = bodyKey !== debounced.key;

  // Carrito vacío → cero, sin arrastrar el último valor cacheado por
  // `keepPreviousData` (si no, al vaciar el carrito quedaría el total viejo).
  return {
    data: hayItems && q.data && !identityPending ? adaptar(q.data) : COTIZACION_VACIA,
    isFetching: q.isFetching || pendingDebounce || identityPending,
    isError: q.isError,
  };
}

/** Etiqueta de la fila de descuento. Cuando gana el del cliente se personaliza
 *  con su nombre ("Descuento para Tincho") — atención manual del dueño.
 *  "manual" (jerarquía, #1219): el admin lo seteó a mano para ESE pedido —
 *  gana outright, no compite con cliente/jornadas. */
export function descuentoLabel(
  origen: DescuentoOrigen,
  jornadas: number,
  nombreCliente?: string | null,
): string {
  if (origen === "manual") {
    return "Descuento manual";
  }
  if (origen === "jornadas") {
    return `Descuento jornadas (${jornadas} ${jornadas === 1 ? "jornada" : "jornadas"})`;
  }
  if (origen === "cliente") {
    const nombre = nombreCliente?.trim();
    return nombre ? `Descuento para ${nombre}` : "Descuento cliente";
  }
  return "";
}
